### PR TITLE
Count vowels- Fixes #23

### LINKED ISSUE #23: Count Vowels in a String
# Count Vowels in a String

## Problem Statement
Write a function `count_vowels(s)` that returns the number of vowels in a given string.

### Examples
- `count_vowels("hello")` should return `2`.
- `count_vowels("sky")` should return `0`.
- `count_vowels("AEIOU")` should return `5`.

## Behavior
The function should:
1. Take a single string `s` as input.
2. Identify and count the vowels (both uppercase and lowercase) in the string.
3. Return the total count of vowels as an integer.

## Resources
- [Python String Methods Documentation](https://docs.python.org/3/library/stdtypes.html#string-methods)
- [Stack Overflow](https://stackoverflow.com/questions/19967001/count-vowels-in-string-python)

### PR DESCRIPTION
**Pull request by fatima-malik99**
---
name: Review solution of count_vowels
about: A template PR for reviewing the code of count_vowels with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by anwar099** [comment]
The count_vowels function and its corresponding unit tests provide a clear and effective solution for counting vowels in a string. The implementation is simple, modular, and easy to understand, making it suitable for real-world usage. Additionally, the unit tests demonstrate thoughtful coverage of a variety of cases, ensuring the function's correctness and robustness. Below is the detailed feedback.

**Strengths**
Clear and concise function and test docstrings.
Covers edge cases (empty strings, no vowels, mixed cases).
Modular and well-structured tests.
**Suggestions**
Use TypeError instead of AssertionError for invalid inputs.
Simplify logic with sum():
return sum(1 for letter in text if letter in "aeiouAEIOU")
Add tests for:
Long strings.
Strings with only whitespace.
Update invalid input test for TypeError.

File: solutions/tests/test_count_vowels.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    - Strings with no vowels
+    - Strings containing only vowels
+    - Mixed-case vowel counting
+    - Alphanumeric strings
+    - Strings with special characters and whitespace
+    - Handling of invalid (non-string) input
+
+Created on: 2025-07-06
+@author: Fatima
+"""
```
Comment (jola-ds) [inline solutions/tests/test_count_vowels.py:12]: Instead of listing the test cases verbatim, consider writing a brief overview including the test categories (like Standard, Edge and Defensive) to keep the docstring more organised.

**Comment by jola-ds** [comment]
Great job, Fatima! 

To keep things more organised, consider rearranging the test cases in a more intuitive order. Start with basic cases, followed by edge cases, and then defensive cases. It looks like it started with edge cases.

I also think the test suite should include one truly basic test, something that does not test edge or specific scenarios, just a typical, basic usage of the function.

File: solutions/count_vowels.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    - count_vowels: Counts the number of vowels in a given string.
+      It supports both uppercase and lowercase vowels.
+
+Created on: 2025-07-06
+@author: Fatima
+"""
+
+
+def count_vowels(text: str) -> int:
+    """Count the number of vowels (a,e,i,o,u) in a string.
```
Comment (jola-ds) [inline solutions/count_vowels.py]: I think ‘’’Returns the number of…’’’is a clearer description of the function’s behaviour because it focuses on the output.

File: solutions/count_vowels.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    >>> count_vowels("why")
+    0
+    """
+    assert isinstance(text, str), "input must be a string"
+    vowels = "aeiouAEIOU"
+    count = 0
+    for letter in text:
+        if letter in vowels:
+            count += 1
+    return count
```
Comment (jola-ds) [inline solutions/count_vowels.py:42]: Consider leaving a space between lines 32 and 33, lines 34 and 35, and lines 37 and 38. This adds proper spacing to separate the different stages of implementation - defensive assertion, variables, loop, and return call

**Review by jola-ds — CHANGES_REQUESTED** [review]
Great job overall! I like how case sensitivity was handled in the implementation. It kept things simple. There are just a few issues (mostly organisation) to address.

File: solutions/count_vowels.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/count_vowels.py:42]: Thank you so much, Jola for pointing that out. I completely ignored that. I'll correct it. 😁

File: solutions/tests/test_count_vowels.py (hunk 1 above)
Comment (fatima-malik99) [reply solutions/tests/test_count_vowels.py:12]: Yes, I was waiting for this because when I was writing this module docstring I was wondering what I should add here. Thank you so much. I'll definitely take this suggestion and improve my docstring.

File: solutions/count_vowels.py (hunk 2 above)
Comment (fatima-malik99) [reply solutions/count_vowels.py]: Hi Jola, that's a good suggestion. But can we write it like this: "Counts the vowels in a string and returns the total?" I think that would be clearer.  Also I just noticed that my function will count both uppercase and lowercase vowels and I've just wrote lowercase vowels so I think that "Returns the number of vowels (both uppercase and lowercase) in a string." would be more better. What's your opinion?

**Comment by fatima-malik99** [comment]
> Great job overall! I like how case sensitivity was handled in the implementation. It kept things simple. There are just a few issues (mostly organisation) to address.

Yes, I just reviewed my code and saw these issues. I really appreciate your efforts for reviewing my code. Thank you!

**Comment by fatima-malik99** [comment]
> Great job, Fatima!
> 
> To keep things more organised, consider rearranging the test cases in a more intuitive order. Start with basic cases, followed by edge cases, and then defensive cases. It looks like it started with edge cases.
> 
> I also think the test suite should include one truly basic test, something that does not test edge or specific scenarios, just a typical, basic usage of the function.

Jola, you are amazing. I will definitely rearrange my tests and ask for rereview again. About your second suggestion, I added edge cases because we are asked in the checklist to add some edge cases. Which test cases do you think are unnecessary? We can have a further discussion on it.

**Comment by fatima-malik99** [comment]
> Jola, you are amazing. I will definitely rearrange my tests and ask for rereview again. About your second suggestion, I added edge cases because we are asked in the checklist to add some edge cases. Which test cases do you think are unnecessary? We can have a further discussion on it.

**Comment by fatima-malik99** [comment]
> Great job, Fatima!
> 
> To keep things more organised, consider rearranging the test cases in a more intuitive order. Start with basic cases, followed by edge cases, and then defensive cases. It looks like it started with edge cases.
> 
> I also think the test suite should include one truly basic test, something that does not test edge or specific scenarios, just a typical, basic usage of the function.

Jola, you are amazing. I will definitely rearrange my tests and ask for rereview again. About your second suggestion, I added edge cases because we are asked in the checklist to add some edge cases. Which test cases do you think are unnecessary? We can have a further discussion on it.

File: solutions/count_vowels.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/count_vowels.py:42]: Hey Jola, I've just tried to add spacing but it's causing Pylint errors.

File: solutions/tests/test_count_vowels.py (hunk 1 above)
Comment (fatima-malik99) [reply solutions/tests/test_count_vowels.py:12]: I've just updated my module docstring. Please check out if this is more suitable. Do tell me if you have any further suggestions.

Comment (jola-ds) [reply solutions/tests/test_count_vowels.py:12]: Well done!! The docstring is organised.

**Comment by fatima-malik99** [comment]
> Great job, Fatima!
> 
> To keep things more organised, consider rearranging the test cases in a more intuitive order. Start with basic cases, followed by edge cases, and then defensive cases. It looks like it started with edge cases.
> 
> I also think the test suite should include one truly basic test, something that does not test edge or specific scenarios, just a typical, basic usage of the function.

Hi I've just rearranged my tests according to your suggestion. I appreciate your feedback in helping me improve the structure! Do tell me if you have further suggestions

File: solutions/tests/test_count_vowels.py (hunk 1 above)
Comment (fatima-malik99) [reply solutions/tests/test_count_vowels.py:12]: Thank you for pointing that out.

File: solutions/count_vowels.py (hunk 2 above)
Comment (jola-ds) [reply solutions/count_vowels.py]: You know, I think that as mentioned in the checklist, the function docstring is supposed to document the behaviour of the function. 

And behaviour simply focuses on the input -> output, what goes in and what comes out. Behaviour does not mention strategy or logic. It's focused on ‘what’ not ‘how’ and I think “count the vowels…” hints at *how* the function works. 

But “returns the number of vowels (both uppercase and lowercase) in a string”  is more focused on the end result, Input is the string, output is the number of vowels. It perfectly reflects function behaviour as a black box. Remember that? 😅 You only see what goes in and what comes out, and you don’t see what goes on in the box.

I think it's just a subtle difference, it doesn’t scream strategy or anything, but I think it can be cleaner or more precise to reflect the behaviour

I’m happy to hear your thoughts on this as well.

Yes! I think specifying uppercase and lowercase makes it super clear 👍

File: solutions/count_vowels.py (hunk 3 above)
Comment (jola-ds) [reply solutions/count_vowels.py:42]: Oh!

Did pylint mention the specific error? It could be a trailing whitespace or something similar.

I suggest formatting the file using ruff or black after adding the spaces. You can run one of the following:

ruff format path\to\solution
Or
Black path\to\soltuion

If you have only ruff installed, that's fine. It should work fine too.

Please try these after adding the spaces and let’s see how that goes.

Comment (fatima-malik99) [reply solutions/count_vowels.py:42]: Yes, Jola. I have just added spacing according to your suggestions using ruff formatting 🤭 Previously, I forgot that we have ruff formatting to save us from Pylint errors 😁. Thank you for reminding me.

**Comment by jola-ds** [comment]
> > Great job, Fatima!
> > To keep things more organised, consider rearranging the test cases in a more intuitive order. Start with basic cases, followed by edge cases, and then defensive cases. It looks like it started with edge cases.
> > I also think the test suite should include one truly basic test, something that does not test edge or specific scenarios, just a typical, basic usage of the function.
> 
> Jola, you are amazing. I will definitely rearrange my tests and ask for rereview again. About your second suggestion, I added edge cases because we are asked in the checklist to add some edge cases. Which test cases do you think are unnecessary? We can have a further discussion on it.

Oh no. I think the edge cases are excellent tests, none is unnecessary.  The test cases perfectly cover the range of the function, they're all necessary. 

My comment was strictly about adding one basic test. Right now, the suite includes edge cases and special cases (which is great!), but it’s missing a truly basic test—one that isn’t about special or edge scenarios, like all_vowels, alphanumeric cases, or repeated vowels. Just one basic test using a typical word would complete the test suite, the test could even be called basic_case. 

You can test any random string containing a vowel, hello, hi,  abcde, or even Fatima! 😅 The most important thing is that the test name and description reflect that it's a basic case.

Thanks for clarifying this!

File: solutions/count_vowels.py (hunk 2 above)
Comment (fatima-malik99) [reply solutions/count_vowels.py]: Actually, you are correct. But when I was writing this docstring I took a hint from the example that Evan shared with us over here: https://github.com/MIT-Emerging-Talent/ET6-Programming-With-Python/blob/main/3_documenting_and_testing/examples/count_vowels.py

He also used Count instead of Returns. But I think that what you are suggesting is totally making sense. Let me correct it. Thank you teamy! 😁

**Comment by jola-ds** [comment]
> > Great job, Fatima!
> > To keep things more organised, consider rearranging the test cases in a more intuitive order. Start with basic cases, followed by edge cases, and then defensive cases. It looks like it started with edge cases.
> > I also think the test suite should include one truly basic test, something that does not test edge or specific scenarios, just a typical, basic usage of the function.
> 
> Hi I've just rearranged my tests according to your suggestion. I appreciate your feedback in helping me improve the structure! Do tell me if you have further suggestions

Well done, Fatima!! Those categories are good additions.

I think no_vowels and empty_string are the only edge cases here, so no_vowels should be moved from the standard category to edge. Repeated_vowels seems more like a standard case, not an edge case because it’s very similar to all_vowels, which is correctly categorized as standard.

The only one I'm unsure about is 
 mixed_case. The reason I think it should be considered standard is the function itself initializes vowels = "aeiouAEIOU", so handling upper and lower cases seems expected. It also fits well with the uppercase_vowels and lowercase_vowels tests, which are standard cases.

What are your thoughts?

**Comment by jola-ds** [comment]
> > Great job overall! I like how case sensitivity was handled in the implementation. It kept things simple. There are just a few issues (mostly organisation) to address.
> 
> Yes, I just reviewed my code and saw these issues. I really appreciate your efforts for reviewing my code. Thank you!

Thanks for the positive engagement and clarifying your thoughts!

File: solutions/count_vowels.py (hunk 2 above)
Comment (jola-ds) [reply solutions/count_vowels.py]: That makes sense! You know, I actually understood this better when Evan used “Returns.…” while solving a problem in class. Someone asked a similar question to yours and Evan’s response clarified things for me too. Also, it really is more of a subtle difference😅

Comment (fatima-malik99) [reply solutions/count_vowels.py]: You are right, the count gives a slight hint about the implementation (i.e., the function counts vowels). While returns focuses on what the function provides as an output. So, return truly treats the function as a black box (input → output). And this is what we have learned. 

So, I am saying yes to Return.  😁

**Comment by fatima-malik99** [comment]
> > > Great job, Fatima!
> > > To keep things more organised, consider rearranging the test cases in a more intuitive order. Start with basic cases, followed by edge cases, and then defensive cases. It looks like it started with edge cases.
> > > I also think the test suite should include one truly basic test, something that does not test edge or specific scenarios, just a typical, basic usage of the function.
> > 
> > 
> > Jola, you are amazing. I will definitely rearrange my tests and ask for rereview again. About your second suggestion, I added edge cases because we are asked in the checklist to add some edge cases. Which test cases do you think are unnecessary? We can have a further discussion on it.
> 
> Oh no. I think the edge cases are excellent tests, none is unnecessary. The test cases perfectly cover the range of the function, they're all necessary.
> 
> My comment was strictly about adding one basic test. Right now, the suite includes edge cases and special cases (which is great!), but it’s missing a truly basic test—one that isn’t about special or edge scenarios, like all_vowels, alphanumeric cases, or repeated vowels. Just one basic test using a typical word would complete the test suite, the test could even be called basic_case.
> 
> You can test any random string containing a vowel, hello, hi, abcde, or even Fatima! 😅 The most important thing is that the test name and description reflect that it's a basic case.
> 
> Thanks for clarifying this!

I think I misunderstood your suggestion previously. Thank you for clarifying. I've added a basic test case name "test_typical_word" . I guess that's what you were talking about. 🧐

**Comment by fatima-malik99** [comment]
> Strings with only whitespace.

Thank you so much for your review. I didn't use TypeError in my test case because, in my code, when I provided an incorrect input, it raised an AssertionError instead of a TypeError.Yes, you are correct, we could use the sum method, but while practicing, I preferred not to because I wanted the code to be simpler, especially when using a debugger, we can't see what's happening inside the code as clearly. But I will definitely add test cases for long strings and strings containing only spaces.
Thank you so much for taking the time to review my code.

**Comment by fatima-malik99** [comment]
> The count_vowels function and its corresponding unit tests provide a clear and effective solution for counting vowels in a string. The implementation is simple, modular, and easy to understand, making it suitable for real-world usage. Additionally, the unit tests demonstrate thoughtful coverage of a variety of cases, ensuring the function's correctness and robustness. Below is the detailed feedback.
> 
> **Strengths** Clear and concise function and test docstrings. Covers edge cases (empty strings, no vowels, mixed cases). Modular and well-structured tests. **Suggestions** Use TypeError instead of AssertionError for invalid inputs. Simplify logic with sum(): return sum(1 for letter in text if letter in "aeiouAEIOU") Add tests for: Long strings. Strings with only whitespace. Update invalid input test for TypeError.

Hi Anwar,
I have added tests for only whitespace and long strings. Thankyou for reviewing.

**Comment by fatima-malik99** [comment]
> > > Great job, Fatima!
> > > To keep things more organised, consider rearranging the test cases in a more intuitive order. Start with basic cases, followed by edge cases, and then defensive cases. It looks like it started with edge cases.
> > > I also think the test suite should include one truly basic test, something that does not test edge or specific scenarios, just a typical, basic usage of the function.
> > 
> > 
> > Hi I've just rearranged my tests according to your suggestion. I appreciate your feedback in helping me improve the structure! Do tell me if you have further suggestions
> 
> Well done, Fatima!! Those categories are good additions.
> 
> I think no_vowels and empty_string are the only edge cases here, so no_vowels should be moved from the standard category to edge. Repeated_vowels seems more like a standard case, not an edge case because it’s very similar to all_vowels, which is correctly categorized as standard.
> 
> The only one I'm unsure about is mixed_case. The reason I think it should be considered standard is the function itself initializes vowels = "aeiouAEIOU", so handling upper and lower cases seems expected. It also fits well with the uppercase_vowels and lowercase_vowels tests, which are standard cases.
> 
> What are your thoughts?

Hey! I am actually surprised by the way you are reviewing my code. I literally just notice that test_all_vowels and test_mixed_case_vowels are both same. Also, you are right no_vowels is an edge case.
I think that I will remove test_all_vowels and keep test_mixed_case inside standard category. Also, I will move test_repeated_vowels to standard category.

**Comment by fatima-malik99** [comment]
> > > > Great job, Fatima!
> > > > To keep things more organised, consider rearranging the test cases in a more intuitive order. Start with basic cases, followed by edge cases, and then defensive cases. It looks like it started with edge cases.
> > > > I also think the test suite should include one truly basic test, something that does not test edge or specific scenarios, just a typical, basic usage of the function.
> > > 
> > > 
> > > Hi I've just rearranged my tests according to your suggestion. I appreciate your feedback in helping me improve the structure! Do tell me if you have further suggestions
> > 
> > 
> > Well done, Fatima!! Those categories are good additions.
> > I think no_vowels and empty_string are the only edge cases here, so no_vowels should be moved from the standard category to edge. Repeated_vowels seems more like a standard case, not an edge case because it’s very similar to all_vowels, which is correctly categorized as standard.
> > The only one I'm unsure about is mixed_case. The reason I think it should be considered standard is the function itself initializes vowels = "aeiouAEIOU", so handling upper and lower cases seems expected. It also fits well with the uppercase_vowels and lowercase_vowels tests, which are standard cases.
> > What are your thoughts?
> 
> Hey! I am actually surprised by the way you are reviewing my code. I literally just notice that test_all_vowels and test_mixed_case_vowels are both same. Also, you are right no_vowels is an edge case. I think that I will remove test_all_vowels and keep test_mixed_case inside standard category. Also, I will move test_repeated_vowels to standard category.

I have added 2 more edge cases as well, test_only_whitespace and test_long_string. What do you think about these?

File: solutions/count_vowels.py (hunk 2 above)
Comment (jola-ds) [reply solutions/count_vowels.py]: Perfect!

**Comment by jola-ds** [comment]
The test suite is perfect! Great job, Fatima!

**Comment by fatima-malik99** [comment]
> The test suite is perfect! Great job, Fatima!

Thank you so much, Jola for reviewing and giving these suggestions.

**Review by jola-ds — APPROVED** [review]

**Comment by jola-ds** [comment]
> > The test suite is perfect! Great job, Fatima!
> 
> Thank you so much, Jola for reviewing and giving these suggestions.

This was great, thanks! 😊
